pkg/controllers/jobset_controller.go

```diff
@@ -15,45 +15,299 @@ package controllers
+// categorized by status (active, successful, failed).
+func (r *JobSetReconciler) getChildJobs(ctx context.Context, jobSet *jobsetv1alpha.JobSet, req ctrl.Request) (*childJobs, error) {
+	// Get all active jobs owned by JobSet.
+	var childJobList batchv1.JobList
+	if err := r.List(ctx, &childJobList, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		klog.Error(err, "unable to list child Jobs")
+		return nil, err
+	}
+
+	jobs := childJobs{}
+	for i, job := range childJobList.Items {
+		_, finishedType := IsJobFinished(&job)
+		switch finishedType {
+		case "": // active
+			jobs.active = append(jobs.active, &childJobList.Items[i])
+		case batchv1.JobFailed:
+			jobs.failed = append(jobs.failed, &childJobList.Items[i])
+		case batchv1.JobComplete:
+			jobs.successful = append(jobs.successful, &childJobList.Items[i])
+		}
+	}
+
+	return &jobs, nil
+}
+
+func (r *JobSetReconciler) createReadyJobs(ctx context.Context, req ctrl.Request, jobSet *jobsetv1alpha.JobSet) error {
+	for _, jobTemplate := range jobSet.Spec.Jobs {
+		job, err := r.constructJobFromTemplate(jobSet, &jobTemplate)
+		if err != nil {
+			klog.Error(err, "unable to construct job from template", "jobTemplate", jobTemplate)
+			return err
+		}
+
+		// Skip the job if it is already active or succeeded.
+		skip, err := r.shouldSkipJob(ctx, job)
+		if err != nil {
+			return err
+		}
+		if skip {
+			klog.Infof("skipping job %s", job.Name)
+			continue
+		}
+
+		klog.Infof("creating job %s", job.Name)
+
+		// Job must be created before headless service so that the job.spec.selector
+		// field is populated.
+		if err := r.Create(ctx, job); err != nil {
+			klog.Error(err, "unable to create Job for JobSet", "job", job)
+			return err
+		}
+
+		// Next, create headless service if specified for this job.
+		if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+			if err := r.createHeadlessSvcIfNotExist(ctx, req, jobSet, job); err != nil {
+				klog.Infof("error creating headless service: %v", err)
+				return err
+			}
+		}
+
+		klog.Info("created Job for JobSet", "job", job)
+	}
+	return nil
+}
+
+// TODO: look into adopting service and updating the selector
+// if it is not matching the job selector.
+func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, req ctrl.Request, jobSet *jobsetv1alpha.JobSet, job *batchv1.Job) error {
+	// Check if service already exists. Service name is same as job name.
+	var headlessSvc corev1.Service
+	namespacedSvc := types.NamespacedName{Namespace: jobSet.Namespace, Name: job.Name}
+
+	if err := r.Get(ctx, namespacedSvc, &headlessSvc); err != nil {
+
+		// Get updated job object from apiserver which will have the selector populated.
+		namespacedJob := types.NamespacedName{Name: job.Name, Namespace: jobSet.Namespace}
+
+		// Need to retry in case the freshly created job object has not been persisted quite yet.
+		if err := retry.OnError(
```

Comment:
Ok makes sense, I updated it to generate our own labels + selectors as described above. Had to use `.` instead of `/` though since the `/` character is not allowed apparently.